### ISSUE TITLE
"detect" command and detection names

### ISSUE THREAD

**Issue by Myzhar** [pjreddie/darknet#335]
I noticed that the command "detect" always read "data/coco.names" even if my "cfg" files points to another file.
"detector test" uses the correct file instead.

**Comment by haemi** [comment]
did you ever find a solution for this?

**Comment by Myzhar** [comment]
Yes, you must use "darknet detector test" so you can set your own "data", "cfg" and "weight"

e.g. 
```$ $ ./darknet detector test data/sea-test-set-v2/trainingSet.data cfg/sea-obstacles-tiny-test_3C-480.cfg 20180208_backup-10.7avg-tiny-COCO-stp-dataset-3C-exp_poly4-Acc_68.510/sea-obstacles-tiny-train_3C-poly_BEST-epochs_530.weights data/sea-test-set-v2/0000.jpg```

**Comment by haemi** [comment]
I tried that:

`./darknet detect test cfg/voc.data cfg/tiny-yolo-voc.cfg backup/tiny-yolo-voc.backup 151229.jpg`

darknet still recognizes stuff like "people" although those objects are not on that image and although my classes do NOT contain such words. Also: I tried to delete the examples folder than to be sure that nothing from the examples is used (e.g. it's classes-file) but now when running the above command I get

`Cannot load image "data/labels/32_0.png"`

seems as if that file is hardcoded somewhere...

**Comment by Myzhar** [comment]
In "voc.data" there is a link to the "names" file "data/voc.names". Be sure that the relative path is correct for you.

When you train your own detector you need to create the correct "label file" and let your "data file" to point to it.

If you want a "tagger" that makes the right job for you, you can use mine:
https://github.com/Myzhar/object-detection-dataset-tagger

It generates the configuration with the relative path to a main folder, usually the one that contains the "darknet" executable

**Comment by haemi** [comment]
@Myzhar thanks a LOT, that really helped!

**Comment by Myzhar** [comment]
You're welcome :+1:

**Comment by adhirmndl** [comment]
Cannot load image "data/labels/32_0.png"
STB Reason: can't fopen

**Comment by illegalplumbing** [comment]
data/labels in the Darknet home folder contains ascii images (32_0.png is the space), and its path is hardcoded into the binary. Copy the folder over to wherever you are running the program from

**Comment by banderlog** [comment]
> data/labels in the Darknet home folder contains ascii images (32_0.png is the space), and its path is hardcoded into the binary. Copy the folder over to wherever you are running the program from

But why it is hardcoded and whats so important in that ascii images?

**Comment by PhilipMachineLearning** [comment]
This is a terrible way to code things - why would you hard core the labels folder?

**Comment by Stern-Hayat** [comment]
Hi.
Someone, please tell me why the error "Not an option: detector train /Users/..../XXX.data" when I attempt to train image data.

**Comment by canseloguzz** [comment]
Hello when i try detector test ;

names: Using default 'data/names.list'
Couldn't open file: data/names.list
 i know i must chance to path, but where no idea.. thanks

**Comment by krisjace** [comment]
I need to be able to use the darknet from any working directory, so I can do scripting. I hacked the darknet executable but it's a very simple and unserious hack. Still, if you screw things up trying to repeat my method, you're on your own.

I did this on Ubuntu 20.04, my darknet home directory is at: **/opt/YOLO/v3/darknet/**
you can see the hard-coded path inside the executable like this:

`$ cat ../darknet | grep -a "labels/"`

and the output will be something like:

> OriginalGrayC1C2C3C4data/labels/%d_%d.pngn >= 0 && n < (int) (sizeof(stbi__bmask)/sizeof(*stbi__bmask))z->code_buffer < (1U << z->num_bits)req_comp >= 1 && req_comp <= 4(((j->code_buffer) >> (32 - h->size[c])) & stbi__bmask[h->size[c]]) == h->code[c]invalid decoded scanline lengthout_n == s->img_n || out_n == s->img_n+1EXPOSURE=          1.0000000000000

so I went ahead to change the hard-coded path directly inside of the binary file using mcedit (sudo apt install mc) as an editor.
If you never alter executables before, just remember that if you change any ASCI string inside an executable you MUST NOT change the total length of the string. You may change character inside the string, but you MUST NOT change the string to a shorter or a longer one than the original.

So I went ahead and linked the data folder under a new symbolic link with a name that was free (you can use a different path as long as you read the information above carefully and keep the original total length of the string):

```
$ sudo ln -s /opt/YOLO/v3/darknet/data /ata
$ sudo mcedit /opt/YOLO/v3/darknet/darknet

```
Then I use the search menu in mcedit to find the occurence of "/labels/" in the binary (there is only 1 occurence in my executable) and manually replace "d" with "/" so that "data/labels" becomes "/ata/labels", save changes, close mcedit.
Now if I run the modified darknet, it uses the absolute path "/ata/labels/" instead of the relative one, and it works like a charm.
You might want to backup the original executable before you edit it.
